// soul-sifter/mixin-alerts.js
import { dedupingMixin } from '@polymer/polymer/lib/utils/mixin.js';

let alertsMixin = (superClass) => class extends superClass {

  static get properties() {
    return {
      alerts: { type: Array }
    };
  }

  constructor() {
    super();
    this._alertsListener = (e) => this._alertsChanged(e);
    this._alertIdPool = 0;
    this.alerts = [];
  }

  connectedCallback() {
    super.connectedCallback();
    window.addEventListener('alerts-changed', this._alertsListener);
  }

  disconnectedCallback() {
    window.removeEventListener('alerts-changed', this._alertsListener);
    super.disconnectedCallback();
  }

  changeAlerts(x) {
    let event = new CustomEvent('alerts-changed', { detail: x });
    window.dispatchEvent(event);
  }

  _alertsChanged(e) {
    this.alertsChanged(e.detail);
  }

  alertsChanged(x) {
    this.alerts = x;
  }

  addAlert(a, timeoutInSeconds = 0) {
    this.alerts.push([++this._alertIdPool, a]);
    this.changeAlerts(this.alerts);
    if (timeoutInSeconds) setTimeout(() => this.rmAlert(this._alertIdPool), timeoutInSeconds * 1000);
    return this._alertIdPool;
  }

  rmAlert(id = undefined) {
    if (id) {
      let idx = this.alerts.findIndex(a => a[0] == id);
      this.alerts.splice(idx, 1);
    } else {
      this.alerts.shift();
    }
    this.changeAlerts(this.alerts);
  }
}

export const AlertsMixin = dedupingMixin(alertsMixin);
